gnome-initial-setup/pages/site/gis-site-search-entry.c

```diff
@@ -0,0 +1,570 @@
+                          G_PARAM_READWRITE));
+
+  g_object_class_install_property (
+      object_class, PROP_SITE,
+      g_param_spec_object (GIS_SITE_SEARCH_ENTRY_PROP_SITE,
+                           "Site",
+                           "The selected site",
+                           GIS_TYPE_SITE,
+                           G_PARAM_READWRITE));
+}
+
+static void
+set_property_sites_available (GisSiteSearchEntry *self,
+                              const GValue       *value)
+{
+  GisSiteSearchEntryPrivate *priv = gis_site_search_entry_get_instance_private (
+    self);
+  GtkListStore *store = NULL;
+  GtkEntryCompletion *completion;
+
+  store = gtk_list_store_new (3, G_TYPE_STRING, GIS_TYPE_SITE, G_TYPE_STRING);
+
+  priv->sites = g_value_dup_boxed (value);
+  for(int i = 0; priv->sites && i < priv->sites->len; i++)
+    {
+      GisSite *site = g_ptr_array_index (priv->sites, i);
+      fill_store (site, store);
+    }
+
+  priv->model = GTK_TREE_MODEL (store);
+  completion = gtk_entry_get_completion (GTK_ENTRY (self));
+  gtk_entry_completion_set_match_func (completion, matcher, NULL, NULL);
+  gtk_entry_completion_set_model (completion, GTK_TREE_MODEL (store));
+}
+
+static void
+set_property (GObject *object, guint prop_id,
+              const GValue *value, GParamSpec *pspec)
+{
+  GisSiteSearchEntryPrivate *priv = gis_site_search_entry_get_instance_private (
+    GIS_SITE_SEARCH_ENTRY (object));
+
+  switch (prop_id)
+    {
+      case PROP_SITES_AVAILABLE:
+        set_property_sites_available (GIS_SITE_SEARCH_ENTRY (object), value);
+        break;
+      case PROP_SITE:
+        gis_site_search_entry_set_site (GIS_SITE_SEARCH_ENTRY (object),
+                                        g_value_dup_object (value));
+        break;
+      default:
+        G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+        break;
+    }
+}
+
+static void
+get_property (GObject *object, guint prop_id,
+              GValue *value, GParamSpec *pspec)
+{
+  GisSiteSearchEntryPrivate *priv = gis_site_search_entry_get_instance_private (
+    GIS_SITE_SEARCH_ENTRY (object));
+
+  switch (prop_id)
+    {
+      case PROP_SITE:
+        g_value_set_boxed (value, priv->site);
+        break;
+      default:
+        G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+        break;
+    }
+}
+
+static void
+entry_changed (GisSiteSearchEntry *entry)
+{
+  GisSiteSearchEntryPrivate *priv =
+      gis_site_search_entry_get_instance_private (entry);
+  GtkEntryCompletion *completion;
+  const gchar *text;
+
+  completion = gtk_entry_get_completion (GTK_ENTRY (entry));
+
+  gtk_entry_completion_set_match_func (gtk_entry_get_completion (GTK_ENTRY (entry)), matcher, NULL, NULL);
+  gtk_entry_completion_set_model (gtk_entry_get_completion (GTK_ENTRY (entry)), priv->model);
+
+  text = gtk_entry_get_text (GTK_ENTRY (entry));
+  if (!text || !*text)
+    set_site_internal (entry, NULL, NULL, NULL);
```

Comment:
I don't think it will work to call `set_site_internal()` with a `NULL` model. But that function should be able to fetch it from the entry on its own already...